stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb

```diff
@@ -1089,6 +1089,15 @@ struct PersistentState {
   static var runNoTestsWasCalled: Bool = false
   static var ranSomething: Bool = false
   static var complaintInstalled = false
+  static var hashingKeyOverridden = false
+
+  static func overrideHashingKey() {
+    if !hashingKeyOverridden {
+      // FIXME(hasher): This has to run before creating the first Set/Dictionary
+      _Hashing.secretKey = (0, 0)
```

Comment:
Obviously, we need some way to set a deterministic key for testing purposes. I'm not expert in this area by miles, but is it preferable to expose a public, undocumented API which a nefarious third-party could conceivably set at will, or would it be safer to make this, say, a build option for the compiler (say, debug only)? (I genuinely don't know.) I recognize that you didn't create this property originally, but now this clearly "does more."

Reply:
Another option to consider is to store a copy of the key in every hash table instance. (That may be worth doing anyway to explicitly prepare for the possibility of per-instance keys.) Doing this would eliminate the awful effects of someone changing the master key while a hash table already exists, even if we keep the key accessible.

Reply:
In pure Swift, AFAIK we don't have a way to guarantee that we could read that environment variable and cache its value before any other Swift code has executed. Can we play some static initialization tricks with the C portion of the standard library to do that instead? Like @xwu I understand the need for this behavior but the possibilities for abuse do worry me.

Reply:
The keys are initialized the runtime before `main()`, in a C++ global static constructor. The code currently uses C++'s Mersenne Twister implementation to generate the two halves of the key. Although we can't run Swift code there, but we could easily call `getenv` or look up some special symbol, or pretty much do whatever we want:

https://github.com/apple/swift/blob/master/stdlib/public/stubs/GlobalObjects.cpp#L109-L120

We could also make use of `HashingDetail.fixedSeedOverride`, if we want to keep the original key around for some reason. (The fixed seed override is currently unused.)

It's unlikely we can protect the key from sufficiently motivated Swift code running within the process; thankfully we don't really need to do that -- the key's role is merely to protect hash table performance against data that is fed into the process from external sources.

Reply:
Eww, is there any alternative to static constructors? As importantly, does that have a corresponding static destructor?

Reply:
Not that I know of! Do you foresee issues?

We don't have destructors, but these are integer values -- AFAIK, we don't have to run any cleanup for them at exit.